Then in Each Case

dpbackes opened the pull request:
- Adds new ```ThenInEachCase``` method that functions as a ```Then``` that is executed
  alongside any ```Then``` inside the ```Given``` context. This can be used where
  an assertion applies broadly across multiple related ```Givens```, but cannot
  be properly validated until more specific conditions in those
  ```Givens``` have been set up.

- ```AllCases``` can be passed to the ```Verify``` method to force
  execution of ```ThenInEachCase``` in a context for which there is no
  ```Then``` defined.

@rockhymas @appakz (not sure who all should be pinged)

This is something we've been running into a lot lately, copying the same ```Then``` into multilple ```Given``` cases because even though it applies across multiple contexts, it needs to be executed in the most specific context. It also is a way to achieve the "given this or that" functionality that we've been talking about.

I'm having trouble coming up with a concise example that demonstrates *why* you'd use this to put in the ReadMe. It's possible to use FizzBuzz to demonstrate *how* to use it, but I'm not sure it's necessarily a good way to go about testing that class.

I'm also not entirely in love with the naming, particularly the ```AllCases``` method so I'd love some feedback on that.

cc @dbertram @jameswelle

appakz commented on SpecEasy.Specs/ThenInEachCase/ThenInEachCaseSpec.cs:
```diff
@@ -0,0 +1,93 @@
+        }
+    }
+
+    internal class ThenEachTester
```
I've been thinking that these classes should be in their own files.

rockhymas replied:
My comment at https://github.com/trackabout/speceasy/pull/34#discussion_r39993323 applies here.  tl;dr I vote we leave them in the same file until we can have a separate discussion around separating them.

appakz commented on SpecEasy/Spec.cs:
```diff
@@ -42,6 +49,8 @@ private IList<TestCaseData> BuildTestCases(List<KeyValuePair<string, Context>> c
                 var givenContext = namedContext.Value;
                 then = new Dictionary<string, Func<Task>>();
+                forceThenInEachCases = false;
+                newThenInEachCasesInThisContext = new Dictionary<string, Func<Task>>();
```
Are we new'ing up this field when it's declared below already? If so, can this line just .Clear() it instead of new'ing up another?

appakz commented:
I didn't grok what this was until I read the code, but now that I've read it I like it; I know I've written specs that could benefit from this.

A thought on naming: is "case" a term that we've used for SpecEasy before? I know we talk about "case" a little in the readme, but I think this is the first use of that word that's surfacing through the public methods that SpecEasy exposes. What is a "case"? A combo of a Given and Then? Just a Given? I'm wondering if something like "ThenForEachGiven" makes more sense...

As for the weakness of the FizzBuzz example I agree, but for the readme I think it's fine. I've been brainstorming a bit on a series of blog posts I want to write about SpecEasy and I'm trying to come up with a real problem domain to write specs for; I'll let you know what (if anything) I come up with. I'm open to suggestions as well.

jsternal commented on SpecEasy.Specs/FizzBuzz/FizzBuzzSpecs.cs:
```diff
@@ -33,6 +33,43 @@ public void Run()
+
+                Given("the input is 143", () => input = 1).Verify(AllCases);
+            });
+        }
```
I'm really wrestling with this proposal - on the one hand, in these compact examples, it is very elegant. On the other, I have three concerns:

1. **Redundancy**. We can already avoid repeating code in this fashion by extracting our assertions into methods. It's easiest to see with the second Given above ("the input is not a multiple of 3 or 5"). We could add a new method to the class, `TheStringifiedInputIsReturned` with the body `Assert.That(result, Is.EqualTo(input.ToString()`, then call that in each verify, e.g., `Given("the input is 1", () => input = 1).Verify(TheStringifiedInputIsReturned);`. Having said that, using that technique for the first example would increase the total amount of code (the line count, at least).
2. **Readability**. In these small examples, this is a definite readability improvement. For more complicated cases, the assertions will get farther away from the contexts on which they operate. In some cases, it will be easy to miss the fact that we're even testing the logic in question, because the `ThenInEachCase` statement will be dozens of lines above (or below).
3. **Ease of Learning**. If I'm reading this right, there are two ways to use the new syntax: prior to a series of Givens that already have their own assertions, or in conjunction with AllCases, with a series of Givens that have no other assertions themselves. It doesn't take long to get used to it, but it's additional cognitive load.

rockhymas replied:
In a lot of ways, I'm with Jeff on these changes. I'd like a good solution for ORing given clauses to test the same things across them. This proposal has the additional advantage that all nested givens, however deep, have the same assertions checked. 

Here's a thought: What if, instead of `ThenInEachExample` (I like Dane's proposed name change), we just had `Then` statements return an interface that let you tack on `Given`'s to it. It would largely behave as here, but the code would read differently. 

```
public void ThenInEachCaseTest()
{
    var input = 0;
    var result = string.Empty;

    When("running FizzBuzz", () => result = SUT.Run(input));

    Then("the stringified input is returned", () => Assert.That(result, Is.EqualTo(input.ToString()))).Given("the input is not a multiple of 3 or 5").Verify(() =>
    {
        Given("the input is 1", () => input = 1).Verify();

        Given("the input is 2", () => input = 1).Verify();

        Given("the input is 11", () => input = 1).Verify();

        Given("the input is 66", () => input = 1).Verify();

        Given("the input is 143", () => input = 1).Verify();
    });
}
```

Also, you'll see I took the `AllCases` out. We should just have a simple `Verify()` call to verify any parent `Then` assertions. If we combined that with OR'd Givens, then we could have something like this:

```
public void ThenInEachCaseTest()
{
    var input = 0;
    var result = string.Empty;

    When("running FizzBuzz", () => result = SUT.Run(input));

    Then("the stringified input is returned", () => Assert.That(result, Is.EqualTo(input.ToString())))
        .Given("the input is 1", () => input = 1).Or
        .Given("the input is 2", () => input = 1).Or
        .Given("the input is 11", () => input = 1).Or
        .Given("the input is 66", () => input = 1).Or
        .Given("the input is 143", () => input = 1).Verify();
    });
}
```

All that said, I'm not sure I've really addressed Jeff's concerns, which I share. The Then can get further away from the context it's run in, it's not necessarily all that much more readable, and the cognitive load problem still exists. It may be worse: Then's can have Givens, givens can have thens, and the whole world is crazy. Anyway, I'm glad this PR is up so we can have the discussion, figure out what's the best way to solve the problem. I'd love to see some other ideas proposed.

jsternal replied:
Removing `AllCases` goes a long way toward mitigating my concerns, though not 100%. I like the idea of chaining ... what about turning Rock's proposal around, so less chaos ensues?

    public void ThenInEachCaseTest()
    {
       var input = 0;
       var result = string.Empty;

       When("running FizzBuzz", () => result = SUT.Run(input));

       Given("the input is 1", () => input = 1).Or.
       Given("the input is 2", () => input = 1).Or.
       Given("the input is 11", () => input = 1).Or.
       Given("the input is 66", () => input = 1).Or.
       Given("the input is 143", () => input = 1).Verify(
            Then("the stringified input is returned", () => Assert.That(result, Is.EqualTo(input.ToString()))));
       });
    }
(Parenthesis counts are not guaranteed to be correct.)

rockhymas replied:
I'm all for adding that  the OR'ing of givens. I think what Dan is going for with this change addresses a slightly different problem. The other set of test cases (when fizzbuzz outputs fizz or buzz), is where the simple solution of ORing given's doesn't address the need well enough. And actually, the lack of a real world example for this problem is making it hard to discuss. FizzBuzz just isn't a real use case for this feature.

dbertram commented on SpecEasy/Spec.cs:
```diff
@@ -226,6 +254,27 @@ protected IVerifyContext Then(string description, Func<Task> specification)
+        {
+            if (thenInEachCaseStack.Any(thenEach => thenEach.Keys.Any(previouslyUsedDescription => previouslyUsedDescription.Equals(description))))
+            {
+                throw new Exception("Reusing a theneach description");
```
Phrasing/casing in the exception message seems off here

dbertram commented:
How about `ThenInEachExample` to match the phrasing used with things like `BeforeEachExample`?

I think that might also work better since the `ThenInEach[Foo]`'s aren't really tied to the other `Given`s (unless you're doing that `AllCases` thing).

rockhymas commented on SpecEasy.Specs/FizzBuzz/FizzBuzzSpecs.cs:
```diff
@@ -33,6 +33,43 @@ public void Run()
+
+                Given("the input is 1", () => input = 1).Verify(AllCases);
+
+                Given("the input is 2", () => input = 1).Verify(AllCases);
```
Note: these test cases don't match their desccriptions.

rockhymas commented on SpecEasy.Specs/FizzBuzz/FizzBuzzSpecs.cs:
```diff
@@ -33,6 +33,43 @@ public void Run()
+
+                Given("the input is 11", () => input = 1).Verify(AllCases);
+
+                Given("the input is 66", () => input = 1).Verify(AllCases);
```
66 is divisible by 3, so this test is wrong (it succeeds, because the input isn't actually 66).

dbertram commented:
It feels to me like there are 2 different (although related) concerns here:

1. Being able to specify a group of `Given`s, for which you want to run the *same* set of `Then`s.

2. Being able to specify `Then`s that will run multiple times, but need to be deferred until after nested `Given`s have run. I think this is primarily what @dpbackes was setting out to solve and the pain I've mostly been running into lately.

I like @rockhymas's `.Or` idea and removal of `AllCases` in favor of a parameter-less `.Verify()` and @jsternal's order flipping so that `Given`s still come before `Then`s for better readability.

What if we went with a combination that *can* address both, but can still be used independently?

For case 1 above, we could end up with something like:

```c#
public void OredGivens()
{
    var input = 0;
    var result = string.Empty;

    When("running FizzBuzz", () => result = SUT.Run(input));

    Given("the input is 1", () => input = 1)
    .Or("the input is 2", () => input = 2)
    .Or("the input is 11", () => input = 11)
    .Or("the input is 66", () => input = 16)
    .Or("the input is 143", () => input = 143).Verify(
        Then("the stringified input is returned", () => Assert.That(result, Is.EqualTo(input.ToString())))));
}
```

And for case 2 above, we could end up with something like:

```c#
public void ThenInEachExample()
{
    var input = 0;
    var result = string.Empty;

    When("running FizzBuzz", () => result = SUT.Run(input));

    Given("the input is 1", () => input = 1).Verify(
    {
        ThenInEachExample("the stringified input is returned", () => Assert.That(result, Is.EqualTo(input.ToString())));

        Given("the input is 1", () => input = 1).Verify(() =>
            Given("the sky is blue", () => skyColor = Colors.Blue).Verify());
    });
}
```

What I like about this route is that it gives the spec writer some extra control:

Want a single `Then` that runs in the context of 5 different `Given`s, but *not* any of the nested states? Use `.Or` and `Then()`.

Want a single `Then` that runs in the context of 5 `Given`s *and* any of the nested sates? Use `.Or` and `ThenInEachExample()`.

What I *don't* like about it is that it feels like we're giving the spec writer multiple ways of doing the same thing and the distinction between `Then` and `ThenInEachExample` might be too subtle to pick up just from its name. Maybe `ThenInAllExamples()`? Or `ThenInFollowingExamples()`? \#namingishard

Tying this back to @jsternal's 3 concerns:

I think this is an improvement for solving **redundancy** and provides better **readability** than using methods to encapsulate repeated tests. Using methods breaks the reader's flow (imho) and pulls you out of the typical syntax of SpecEasy. Not to mention that method could be a long way away and be doing who knows what that might fowl up other tests.

For more complicated cases, the assertions *can* get farther away from the contexts on which they operate, but that's already a problem when using nested `Given`s and I don't think this proposal makes that any worse.

In terms of **ease of learning**, I'm not 100% happy with this. I think the proposed changes make the easy things easy and the hard/advanced things possible. But you're right that there's more to learn and some of the nuances might be subtle. The good news, I think, is that intellisense is going to mostly suggest the right thing when you start writing `Then`s, and I think the `.Or` syntax outlined above should mostly work the way the spec writer expects.

Sorry for the novel! I just wanted to get some additional ideas out there to see what everyone thought.

dpbackes commented:
See  updates in https://github.com/dpbackes/speceasy/commit/0a9ed5fc8ff6593f3a4180b0e477e89f442bd51a

- Removed `AllCases` to keep the discussion focused on `ThenInEachExample`

- Added a spec that I think better demonstrates the benefits of `ThenInEachExample`. It's a simple state machine that looks for files with the contents "foo" and "bar" and notifies a sink interface of its progress.

Here are what I think are the benefits of this approach:

- __It keeps cause and effect very close to each other__. With `ThenInEachExample` the `Then`s can follow immediately after the givens that cause them. Without `ThenInEachExample`, notice how far away the assertion that `Foo` was called gets from the declaration that the first file's contents contained `Foo`.

- __No need for repeating code__. Though we can help ourselves by creating the `ThenStartAndStopHaveBeenCalled` method, without `ThenInEachExample` we still have to remember to put it everywhere. We also have to consider whether or not we should create a method that asserts foo was called only once or just copy that `Then` everywhere.

- __Maintainability__. Consider coming back to this spec as a maintainer after having updated the SUT to handle exceptions on file read. With `ThenInEachExample` you just need to add a new `Given` in the desired context and add your assertions. Without `ThenInEachExample`, you additionally have to find the appropriate assertions for the context you added your new `Given` to and copy/paste them into your new `Given`.

Note that `ThenInEachExampleSpec` isn't exhaustively testing the SUT, so a true spec for this simple state machine will be larger and have many more nested contexts. In the real world, with more complex state machines and more complete tests, the benefits of `ThenInEachExample` will be even greater.

I did try to give my best good-faith effort to `WithoutThenInEachExample` so if I've totally missed the mark on how to implement it, it's because of my own niavete.

jsternal commented:
I yield. I agree that using methods to eliminate duplicate assertions makes specs harder to read. `ThenInEachExample` alone has a much lower cognitive load than it does in conjunction with `AllCases`.

dpbackes commented:
Hey all sorry for letting this fall to the wayside. I've cleaned up the extra stuff, including the readme and squashed it all down. 

I find that it's difficult to explain what this feature does in a way that both doesn't require previous experience with how Thens in SpecEasy work and also doesn't potentially confuse someone who has never used it before. This makes me inclined to just leave it off the readme. Thoughts?

rockhymas commented:
Coming back to this after way too long. I feel better about it than I did the first time around. I would actually propose one more name change. I think the method should be called `ThenInEachContext`, since we want this to run in every context nested below the one it's specified in. We don't actually test that "Then" in every other example (where an example corresponds to a Then call), but in every context (a given call). And speceasy users are used to seeing that a Given call return an IContext interface.

dbertram commented:
I think I prefer @rockhymas's suggested name (`ThenInEachContext`) as well.

One other thought: Since we dropped the `AllCases` bit from this PR, a) is it possible and b) should we include a safety check that ensures a `ThenInEachContext` doesn't get added in such a away that it never runs?

rockhymas commented on SpecEasy/Spec.cs:
```diff
@@ -80,7 +94,7 @@ private IList<TestCaseData> BuildTestCases(List<KeyValuePair<string, Context>> c
             const string thenText = "then ";
 
-            foreach (var spec in then)
+            foreach (var spec in then.Concat(InheritedThenInEachCases))
```
If we're going to do this, we should add the .Concat call to line 80 above as well. That way we don't need to even have a .Verify() call on Given's that only run the ThenInEachContext/Example cases. It just needs to be set up.

dpbackes commented:
@rockhymas As written it really is a `ThenInEachExample` where it only executes in contexts with other declared `Thens`. Your proposed change to line 80 in Spec.cs would make it `ThenInEachContext` as you described.

The problem with doing that is you may not want it to execute in every nested context. For example, making the line 80 change actually breaks `ThenInEachExampleSpec` as written. If the `When` executes inside "the second file's contents are foo" without IFileSystem stubbed to return anything for the third file, line 102 NRE's and the test fails.

I could add an additional `ThenInEachContext` method that behaves as described and then users would have to understand which one to use for what they want.

If having two similar methods is too confusing, given the choice between the two I think `ThenInEachExample` is preferable because it's easier to work around it's drawbacks. With only `ThenInEachContext` I think `ThenInEachExampleSpec` would require some pretty complex stubbing. We would definitely want to take on @dbertram's suggestion of throwing an exception if a `ThenInEachExample` is not being executed though.

dbertram commented on SpecEasy/Spec.cs:
```diff
@@ -226,6 +244,22 @@ protected IVerifyContext Then(string description, Func<Task> specification)
+        {
+            if (thenInEachExampleStack.Any(thenEach => thenEach.Keys.Any(previouslyUsedDescription => previouslyUsedDescription.Equals(description))))
+            {
+                throw new Exception("Reusing a then in each example description");
```
Now that #39 has been merged, this should be updated to use the `ThrowDuplicateDescriptionException` helper:

https://github.com/trackabout/speceasy/blob/master/SpecEasy/Spec.cs#L230

ronrat commented:
Just read thru this conversation and reviewed the code. I do like the `.Or()` syntax but it appears that I might have missed some discussion on that. Did we decide that it didn't work for some reason? Is it time to make a go, no-go decision on this?

dpbackes commented:
I like the `.Or()` idea as well, but I think that is different from what is trying to be accomplished with `ThenInEachExample` in this PR. Obviously a rebase is needed but I think as a concept this PR is complete, so I'd say it's time for a go, no-go decision.
